### ISSUE TITLE
Data Collector misses tons of resources

### ISSUE THREAD

**Issue by jaymzh** [chef/chef#8049]
## Description
The datacollector code is filled with `unless nested_resource?(resource)`.

Our primary use-case for datacollector is to get the data about child runcontexts we can't access from report handlers or event handlers (see https://github.com/chef/chef/issues/8044).

Please update datacollector to include nested resources, possibly with a config if necessary.

## Chef Version
14.8.12

## Platform Version
CentOS 7

**Comment by lamont-granquist** [comment]
So #8063 is going this direction.

The question I've got is what the JSON output should look like?  Right now in the action collection I've kept it flat and the wrapping resource appears after all the sub-resources.  There's a single integer for how many nesting levels deep the resource is.  That's the minimum viable product here most likely, makes it easy for consumers to just rip through the whole thing and not care about nesting at all as well (optimizing for the case where the user cares that there's a file resource on a given file but has no interest in how many levels deep inside of a resource it occurs).

**Comment by jaymzh** [comment]
I think the _best_ choice would be to do something like 
{
  'name': 'my_resource[dingleberry]',
  'action': ':run',
  ...
  'child_resources': {
    {
      'name': 'template[shoopshoop]',
      'action': ':create',
      ...
    },
    ...
  },
},
```

This would be useful to be able to debug weird problems, to figure out where two duplicate child-resources in different run contexts were coming from, etc...

I think it's not **absolutely** necessary to do it this way - your way will get us what we need at the moment. My concern is backcompat. If we do the less-expressive way now, changing it later is hard.

**Comment by jaymzh** [comment]
And then if we do that, it'd be useful to may have some helpers like:

```
find_all_resources(type)
```
or

```
get_flattened_list_of_resources
```

**Comment by lamont-granquist** [comment]
Yeah I'll have to figure out first if the action_collection itself should be organized flat internally and then has methods to expose it nested, or is nested internally and has methods exposed to make it flat.

**Comment by lamont-granquist** [comment]
closed by #8063
